keras_nlp/layers/transformer_decoder.py

```diff
@@ -233,14 +238,29 @@ def call(
             encoder_attention_mask: a boolean Tensor. Customized encoder
                 sequence mask, must of shape
                 [batch_size, encoder_sequence_length, encoder_sequence_length].
-            cache: a dense float Tensor. The cache of key/value of leading
+            self_attention_cache: a dense float Tensor. The cache of key/value of leading
                 tokens. `cache` is of shape [B, 2, max_seq_len, num_heads,
                 key_dims].
-            cache_index: a int or int Tensor, the index of the current token
+            cache_index: an int or int Tensor, the index of the current token
                 being processed.
+            cross_attention_cache: a dense float Tensor. The cache of
+                key/value pairs of encoder outputs. `cache` is of shape
+                `[B, 2, S, num_heads, key_dims]`.
+            compute_cross_attention_cache: a boolean. Whether to compute the
+                cross-attention cache. This is useful when
+                `cross_attention_cache = None` and
+                `compute_cross_attention_cache = True` for the first decoder
+                forward pass. If `cross_attention_cache` is not `None` and
+                `compute_cross_attention_cache` is `True`, we do not recompute
+                the cache.
         Returns:
-            Either a tuple of (outputs, cache) if a cache was passed, or a
-            single value outputs if a cache was not passed.
+            If the layer does not have a cross-attention layer, returns a tuple
```

Comment:
I'm pretty sure we can simplify the flow of arguments here quite a bit. Here's a draft idea, may need tweaking...

1. Remove `CachedMultiHeadCrossAttention` entirely.
2. Rename `cache_index` to `cache_update_index` in `CachedMultiHeadAttention`. Logic update...
```
        if cache is not None:
            key_cache, value_cache = tf.unstack(cache, axis=1)
            if cache_update_index is not None:
                start = [0, cache_index, 0, 0]
                key = dynamic_update_slice(key_cache, key, start)
                value = dynamic_update_slice(value_cache, value, start)
                cache = tf.stack((key, value), axis=1)
```
3. In this layer, take in `encoder_cache`, `encoder_cache_update_index`, `decoder_cache`, `decoder_cache_update_index`. (remove `compute_cross_attention_cache`)
4. Return either `x` if no cache. `x, decoder_cache` if no cross attention, and `x, encoder_cache, decoder_cache` if cross attention. We can just error if only one cache is set in the cross attention case (there is no use case right?).
5. Seed the cache by calling with `encoder_cache_update_index=0` and `decoder_cache_update_index=0`. Follow up in the loop by calling with `encoder_cache_update_index=None` and `decoder_cache_update_index=i`.

Let me know if you think this would work! And if so, we should do some rough timing tests to make sure this is ballpark the same performance.

But overall I think we need something like this to make this logic a little more easy to follow.

Reply:
From meeting, maybe this logic instead.

```python
if cache is not None:
    key, value = tf.unstack(cache, axis=1)
    if cache_update_index is not None:
        key_update = self._key_dense(key)
        value_update = self._value_dense(value)
        start = [0, cache_index, 0, 0]
        key = dynamic_update_slice(key_cache, key_update, start)
        value = dynamic_update_slice(value_cache, value_update, start)
        cache = tf.stack((key, value), axis=1)
else:
    key = self._key_dense(key)
    value = self._value_dense(value)
```